system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1276,416 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+                                             g_variant_new ("(s)", sender),
+                                             G_VARIANT_TYPE ("(a{sv})"), G_DBUS_CALL_FLAGS_NONE,
+                                             G_MAXINT, NULL, error);
+  if (credentials == NULL)
+    return FALSE;
+
+  dict = g_variant_get_child_value (credentials, 0);
+
+   if (!g_variant_lookup (dict, "UnixUserID", "u", out_uid))
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to query UnixUserID for the bus name: %s", sender);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+check_for_system_helper_user (struct passwd  *passwd,
+                              gchar         **passwd_buf,
+                              GError        **error)
+{
+  struct passwd *result = NULL;
+  g_autofree gchar *buf = NULL;
+  size_t bufsize;
+  int err;
+
+  bufsize = sysconf (_SC_GETPW_R_SIZE_MAX);
+  if (bufsize == -1)          /* Value was indeterminate */
+     bufsize = 16384;        /* Should be more than enough */
+
+  while (!result)
+    {
+      buf = g_malloc0 (bufsize);
+      err = getpwnam_r (SYSTEM_HELPER_USER, passwd, buf, bufsize, &result);
+      if (result == NULL)
+        {
+          if (err == ERANGE)     /* Insufficient buffer space */
+            {
+              g_free (buf);
+              bufsize *= 2;
+              continue;
+            }
+          else if (err == 0)     /* User SYSTEM_HELPER_USER 's record was not found*/
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                           "User %s does not exist in password file entry", SYSTEM_HELPER_USER);
+              return FALSE;
+            }
+          else
+            {
+              g_set_error (error, G_IO_ERROR, g_io_error_from_errno (err),
+                           "Failed to query user %s from password file entry", SYSTEM_HELPER_USER);
+              return FALSE;
+            }
+        }
+    }
+
+  *passwd_buf = g_steal_pointer (&buf);
+
+  return TRUE;
+}
+
+static void
+revokefs_fuse_backend_child_setup (gpointer user_data)
+{
+  struct passwd *passwd = user_data;
+
+  if (setgid (passwd->pw_gid) == -1)
+    g_warning ("Failed to setgid(%d) for revokefs backend: %s",
+               passwd->pw_gid, g_strerror (errno));
+
+  if (setuid (passwd->pw_uid) == -1)
+    g_warning ("Failed to setuid(%d) for revokefs backend: %s",
+               passwd->pw_uid, g_strerror (errno));
+}
+
+static void
+name_vanished_cb (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  const gchar *unique_name = (const gchar *) user_data;
+  OngoingPull *pull;
+
+  G_LOCK (ongoing_pulls);
+  for (guint i = 0; i < ongoing_pulls->len; i++)
+    {
+      pull = g_ptr_array_index (ongoing_pulls, i);
+      if (g_strcmp0 (pull->unique_name, unique_name) == 0)
+        {
+          g_ptr_array_remove (ongoing_pulls, pull);
+          ongoing_pull_free (pull);
```

Comment:
ongoing_pull_free() does a rm -rf which is kinda slow. Maybe we should do the frees outside the lock.

Reply:
If I understand D-bus unique names correctly (I hope), then  this will lose out on the case where there are multiple pulls in the array for the same unique name. We should be free-ing up each and every pull for that connection, right?

Reply:
Yeah, so you would need to remoe the pulls from ongoing_pulls and put them in a local array, then free those outside the lock.

Reply:
Addressed using a separate array and freeing it outside the locks.